src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala

```diff
@@ -55,46 +55,45 @@ private[deequ] case class AnalysisBasedConstraint[S <: State[S], M, V](
       analysisResults: Map[Analyzer[_, Metric[_]], Metric[_]])
     : ConstraintResult = {
 
-    val maybeValue: Option[Try[M]] = analysisResults
-      .get(analyzer)
-      .map { _.value.asInstanceOf[Try[M]] }
-
-    maybeValue
-      .map {
-        // Analysis done successfully and result metric is there
-        case Success(metricValue) => pickValueAndAssert(metricValue)
-        // An exception occurred during analysis
-        case Failure(e) => ConstraintResult(this, ConstraintStatus.Failure, Option(e.getMessage))
-      }
-      .getOrElse(
-        // Analysis is missing
-        ConstraintResult(this, ConstraintStatus.Failure,
-          Some(AnalysisBasedConstraint.MissingAnalysis)))
-  }
-
-  private[this] def pickValueAndAssert(metricValue: M): ConstraintResult = {
-    try {
-      val assertOn = runPickerOnMetric(metricValue)
-      val assertionOk = runAssertion(assertOn)
-
-      if (assertionOk) {
-        ConstraintResult(this, ConstraintStatus.Success)
-      } else {
-        var errorMessage = s"Value: $assertOn does not meet the constraint requirement!"
+    val metric = analysisResults.get(analyzer).map(_.asInstanceOf[Metric[M]])
 
-        hint.foreach(hint => errorMessage += s" $hint")
-
-        ConstraintResult(this, ConstraintStatus.Failure,
-          Some(errorMessage))
-      }
+    metric.map(pickValueAndAssert).getOrElse(
+      // Analysis is missing
+      ConstraintResult(this, ConstraintStatus.Failure,
+        Some(AnalysisBasedConstraint.MissingAnalysis), metric)
+    )
+  }
 
-    } catch {
-      case AnalysisBasedConstraint.ConstraintAssertionException(msg) =>
-        ConstraintResult(this, ConstraintStatus.Failure,
-          Some(s"${AnalysisBasedConstraint.AssertionException}: $msg!"))
-      case AnalysisBasedConstraint.ValuePickerException(msg) =>
-        ConstraintResult(this, ConstraintStatus.Failure,
-          Some(s"${AnalysisBasedConstraint.ProblematicMetricPicker}: $msg!"))
+  private[this] def pickValueAndAssert(metric: Metric[M]): ConstraintResult = {
+
+    metric.value match {
+      // Analysis done successfully and result metric is there
+      case Success(metricValue) =>
+        try {
+          val assertOn = runPickerOnMetric(metricValue)
+          val assertionOk = runAssertion(assertOn)
+
+          if (assertionOk) {
+            ConstraintResult(this, ConstraintStatus.Success, metric = Some(metric))
+          } else {
+            var errorMessage = s"Value: $assertOn does not meet the constraint requirement!"
+
+            hint.foreach(hint => errorMessage += s" $hint")
```

Comment:
use mkString

Reply:
:+1:

Reply:
First thought mkString would make it more readable and compact. However also `None.mkString(" ", "", "")` prints a leading space. I'll keep the foreach instead.